### PR TITLE
Disallow setting init-only property in nested initializer

### PR DESCRIPTION
**Pull request by jcouv**
Fixes https://github.com/dotnet/roslyn/issues/50126

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Generated/BoundNodes.xml.Generated.cs [hunk 1]
```diff
@@ -562,32 +562,36 @@ public BoundDisposableValuePlaceholder Update(TypeSymbol type)
...
             RoslynDebug.Assert(type is object, "Field 'type' cannot be null (make the type nullable in BoundNodes.xml to remove this check)");
 
+            this.IsNew = isNew;
         }
 
 
         public new TypeSymbol Type => base.Type!;
+
+        public bool IsNew { get; }
         [DebuggerStepThrough]
```
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/Generated/BoundNodes.xml.Generated.cs:587]: Unrelated to this PR: Does it worth to update the code that generates this to leave an empty line after this property?

Comment (jcouv) [reply src/Compilers/CSharp/Portable/Generated/BoundNodes.xml.Generated.cs:587]: Meh, we have lots of functional issues. Formatting generated code isn't worth spending time :-/

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 2]
```diff
@@ -4412,6 +4412,7 @@ private BoundExpression MakeBadExpressionForObjectCreation(SyntaxNode node, Type
                 var boundInitializer = BindInitializerExpression(syntax: initializerOpt,
                                                                  type: type,
                                                                  typeSyntax: typeSyntax,
+                                                                 isNew: true,
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:4415]: Is this to suppress cascading errors?

File: src/Compilers/CSharp/Portable/Binder/Binder_WithExpression.cs [hunk 3]
```diff
@@ -54,6 +54,7 @@ private BoundExpression BindWithExpression(WithExpressionSyntax syntax, Diagnost
                 syntax.Initializer,
                 receiverType,
                 syntax.Expression,
+                isNew: true,
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Binder/Binder_WithExpression.cs:57]: >isNew [](start = 16, length = 5)

Maybe `IsNew` isn't the greatest name here. Perhaps something like `IsNewObject` or `IsNewOrWith`? `IsNew` on it's own implies a literal `new` syntax node to me.

File: src/Compilers/CSharp/Test/Semantic/Semantics/InitOnlyMemberTests.cs [hunk 4]
```diff
@@ -4035,5 +4035,152 @@ .maxstack  2
...
+[ComImport()]
+[Guid(""f9c2d51d-4f44-45f0-9eda-c9d599b58277"")]
+[CoClass(typeof(ClassITest28))]
+public interface ITest28
+{
+    int Property { get; init; }
+}
+
+[Guid(""f9c2d51d-4f44-45f0-9eda-c9d599b58278"")]
+public abstract class ClassITest28 //: ITest28
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/InitOnlyMemberTests.cs:4160]: Is this class even necessary?

File: src/Compilers/CSharp/Test/Semantic/Semantics/InitOnlyMemberTests.cs [hunk 5]
```diff
@@ -4035,5 +4035,152 @@ .maxstack  2
...
+            var piaCompilation = CreateCompilationWithMscorlib45(new[] { IsExternalInitTypeDefinition, pia }, options: TestOptions.DebugDll);
+
+            CompileAndVerify(piaCompilation);
+
+            string source = @"
+class UsePia
+{
+    public static void Main()
+    {
+        var x1 = new ITest28();
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/InitOnlyMemberTests.cs:4175]: The property doesn't appear to actually be being used.

**Comment by 333fred** [comment]
Done review pass (commit 2).

File: src/Compilers/CSharp/Test/Semantic/Semantics/InitOnlyMemberTests.cs (hunk 5 above)
Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/Semantics/InitOnlyMemberTests.cs:4175]: Also, do we have a NoPia test doing a legal init assignment?

---
In reply to: [554113056](https://github.com/dotnet/roslyn/pull/50200#discussion_r554113056) [](ancestors = 554113056)

File: src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml [hunk 6]
```diff
@@ -117,6 +117,7 @@
   <!-- The implicit collection in an object or collection initializer expression. -->
   <Node Name="BoundObjectOrCollectionValuePlaceholder" Base="BoundValuePlaceholderBase">
     <Field Name="Type" Type="TypeSymbol" Override="true" Null="disallow"/>
+    <Field Name="IsNew" Type="bool" Null="NotApplicable"/>
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:120]: >IsNew [](start = 17, length = 5)

"IsObjectBeingCreated" or "IsInstanceBeingCreated"?

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 7]
```diff
@@ -4423,12 +4424,13 @@ private BoundExpression MakeBadExpressionForObjectCreation(SyntaxNode node, Type
             InitializerExpressionSyntax syntax,
             TypeSymbol type,
             SyntaxNode typeSyntax,
+            bool isNew,
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:4427]: >isNew [](start = 17, length = 5)

"isForNewInstance"?

File: src/Compilers/CSharp/Test/Semantic/Semantics/InitOnlyMemberTests.cs (hunk 5 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/Semantics/InitOnlyMemberTests.cs:4175]: Yes, it doesn't look like this unit-test actually tests something interesting. 

---
In reply to: [554113056](https://github.com/dotnet/roslyn/pull/50200#discussion_r554113056) [](ancestors = 554113056)

**Comment by AlekseyTs** [comment]
Done with review pass (commit 2)

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 2 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:4415]: This is to differentiate `new C() { $$.P = ... }` and `new C() { P = { $$.P2 = ... }`. The first one is a new instance. The latter is updating an existing instance.

---
In reply to: [554110842](https://github.com/dotnet/roslyn/pull/50200#discussion_r554110842) [](ancestors = 554110842)

Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:4415]: Sorry, I didn't see the comment was in the handling for bad cases. Yes, for error scenarios, it doesn't seem useful to add other errors.

---
In reply to: [554174635](https://github.com/dotnet/roslyn/pull/50200#discussion_r554174635) [](ancestors = 554174635,554110842)

File: src/Compilers/CSharp/Test/Semantic/Semantics/InitOnlyMemberTests.cs (hunk 4 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/InitOnlyMemberTests.cs:4160]: This is an existing test that I copied. I had the same thought and tried removing that type, but that added diagnostics. I didn't spend much time to figure it out...

---
In reply to: [554112878](https://github.com/dotnet/roslyn/pull/50200#discussion_r554112878) [](ancestors = 554112878)

File: src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml (hunk 6 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:120]: How about "IsNewInstance"?

---
In reply to: [554125889](https://github.com/dotnet/roslyn/pull/50200#discussion_r554125889) [](ancestors = 554125889)

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:120]: > How about "IsNewInstance"?

Sounds fine to me.


---
In reply to: [554191156](https://github.com/dotnet/roslyn/pull/50200#discussion_r554191156) [](ancestors = 554191156,554125889)

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 7 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:4427]: Still needs to be addressed.

---
In reply to: [554130634](https://github.com/dotnet/roslyn/pull/50200#discussion_r554130634) [](ancestors = 554130634)

**Comment by 333fred** [comment]
Done review pass (commit 3)

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 7 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: Thanks.
Note: I merged with latest test from master. I made a small modification to report both the langver and invalid consumption of init-only property.

---
In reply to: [554220332](https://github.com/dotnet/roslyn/pull/50200#discussion_r554220332) [](ancestors = 554220332,554130634)

**Review by 333fred — APPROVED** [review]
LGTM (commit 5)

File: src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs [hunk 8]
```diff
@@ -1086,16 +1086,22 @@ private bool CheckPropertyValueKind(SyntaxNode node, BoundExpression expr, BindV
...
                             Error(diagnostics, ErrorCode.ERR_AssignmentInitOnly, node, propertySymbol);
-                            return false;
+                            hasInitError = true;
                         }
 
                         if (setMethod.DeclaringCompilation != this.Compilation)
                         {
                             // an error would have already been reported on declaring an init-only setter
-                            CheckFeatureAvailability(node, MessageID.IDS_FeatureInitOnlySetters, diagnostics);
+                            hasInitError |= !CheckFeatureAvailability(node, MessageID.IDS_FeatureInitOnlySetters, diagnostics);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs:1099]: > hasInitError |= !CheckFeatureAvailability(node, MessageID.IDS_FeatureInitOnlySetters, diagnostics); [](start = 27, length = 100)

This change doesn't feel right to me. Targeting language version shouldn't change how we bind things, it should only affect feature availability diagnostics. Also, I don't think we would want the binding behavior to be different based on where the property is defined. Bottom line, the only difference is acceptable is presence or lack of feature availability diagnostics, nothing else.

File: src/Compilers/CSharp/Test/Semantic/Semantics/InitOnlyMemberTests.cs [hunk 9]
```diff
@@ -113,6 +113,9 @@ void M(C c)
             var comp = CreateCompilation(source, parseOptions: TestOptions.Regular8,
                 references: new[] { useMetadataImage ? lib.EmitToImageReference() : lib.ToMetadataReference() });
             comp.VerifyEmitDiagnostics(
+                // (6,45): error CS8852: Init-only property or indexer 'C.Property' can only be assigned in an object initializer, or on 'this' or 'base' in an instance constructor or an 'init' accessor.
+                //         _ = new Container() { contained = { Property = string.Empty, Property2 = string.Empty } };
+                Diagnostic(ErrorCode.ERR_AssignmentInitOnly, "Property").WithArguments("C.Property").WithLocation(6, 45),
                 // (6,45): error CS8400: Feature 'init-only setters' is not available in C# 8.0. Please use language version 9.0 or greater.
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/InitOnlyMemberTests.cs:119]: >// (6,45): error CS8400: Feature 'init-only setters' is not available in C# 8.0. Please use language version 9.0 or greater. [](start = 16, length = 124)

I am not sure it is worth reporting this error here at all. No matter what version is used, an init-only property cannot-be assigned here. I'd say it is probably better to not make any changes in ```CheckPropertyValueKind```.

File: src/Compilers/CSharp/Test/Semantic/Semantics/InitOnlyMemberTests.cs [hunk 10]
```diff
@@ -188,6 +191,9 @@ void M(C c)
                 // (6,9): error CS8852: Init-only property or indexer 'C.Property' can only be assigned in an object initializer, or on 'this' or 'base' in an instance constructor or an 'init' accessor.
                 //         c.Property = string.Empty;
                 Diagnostic(ErrorCode.ERR_AssignmentInitOnly, "c.Property").WithArguments("C.Property").WithLocation(6, 9),
+                // (6,9): error CS8400: Feature 'init-only setters' is not available in C# 8.0. Please use language version 9.0 or greater.
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/InitOnlyMemberTests.cs:194]: >// (6,9): error CS8400: Feature 'init-only setters' is not available in C# 8.0. Please use language version 9.0 or greater. [](start = 16, length = 123)

Same comment here

**Comment by AlekseyTs** [comment]
Done with review pass (commit 5)

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 6), assuming CI is passing.

**Pull request merged**